addons/webhooks/clusterbootstrap_webhook.go

```diff
@@ -0,0 +1,413 @@
+
+	if oldPkg == nil || newPkg == nil {
+		return field.Invalid(fldPath, newPkg, "package can't be nil")
+	}
+
+	// Enforce version and refName check for packages. We won’t allow packageRef’s to be downgraded or change the package from something like calico to antrea
+	// both old and new packages should be present in the cluster
+	newPackageRefName, newPackageVersion, err := util.GetPackageMetadata(ctx, wh.aggregatedAPIResourcesClient, newPkg.RefName, wh.SystemNamespace)
+	if err != nil {
+		return field.Invalid(fldPath.Child("refName"), newPkg.RefName, err.Error())
+	}
+
+	oldPackageRefName, oldPackageVersion, err := util.GetPackageMetadata(ctx, wh.aggregatedAPIResourcesClient, oldPkg.RefName, wh.SystemNamespace)
+	if err != nil {
+		return field.Invalid(fldPath.Child("refName"), oldPkg.RefName, err.Error())
+	}
+
+	// RefName within the package CR should stay the same
+	// For core packages, an example would be user can't switch the CNI from Antrea to Calico
+	// For additional packages, the package can't be removed once added
+	if newPackageRefName != oldPackageRefName {
+		return field.Invalid(fldPath.Child("refName"), newPkg.RefName, "new package refName and old package refName should be the same")
+	}
+
+	// The package can't be downgraded
+	newPkgSemver, err := versions.NewRelaxedSemver(newPackageVersion)
+	if err != nil {
+		retErr := errors.Wrap(err, "new package version is invalid")
+		return field.Invalid(fldPath.Child("refName"), newPkg.RefName, retErr.Error())
+	}
+	oldPkgSemver, err := versions.NewRelaxedSemver(oldPackageVersion)
+	if err != nil {
+		retErr := errors.Wrap(err, "old package version is invalid")
+		return field.Invalid(fldPath.Child("refName"), oldPkg.RefName, retErr.Error())
+	}
+	if newPkgSemver.Compare(oldPkgSemver.Version) == -1 {
+		// package downgrade is not allowed
+		return field.Invalid(fldPath.Child("refName"), newPkg.RefName, "package downgrade is not allowed")
+	}
+
+	if err := wh.validateValuesFromUpdate(ctx, oldPkg.ValuesFrom, newPkg.ValuesFrom, fldPath.Child("valuesFrom")); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// validateValuesFromUpdate validates contents of valuesFrom in upgrade
+func (wh *ClusterBootstrap) validateValuesFromUpdate(_ context.Context, oldValuesFrom, newValuesFrom *runv1alpha3.ValuesFrom, fldPath *field.Path) *field.Error {
+	if oldValuesFrom != nil && newValuesFrom != nil {
+		// We don't allow changes to APIGroup and Kind of providerRef
+		if oldValuesFrom.ProviderRef != nil && newValuesFrom.ProviderRef != nil {
+			if *oldValuesFrom.ProviderRef.APIGroup != *newValuesFrom.ProviderRef.APIGroup ||
+				oldValuesFrom.ProviderRef.Kind != newValuesFrom.ProviderRef.Kind {
+				return field.Invalid(fldPath.Child("ProviderRef"), newValuesFrom.ProviderRef, "change to Group and Kind in ProviderRef is not allowed")
+			}
+		}
+
+		// Users can't switch from ProviderRef to secretRef/inline
+		if oldValuesFrom.ProviderRef != nil && newValuesFrom.ProviderRef == nil {
```

Comment:
This is checking if new ProviderRef is nil and not what the comment states.

Reply:
Updated